src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,498 @@
+            {
+                return false;
+            }
+
+            kind = (UnsafeAccessorKind)decodedValue.FixedArguments[0].Value!;
+
+            // Check the name of the target to access. This is the name we
+            // use to look up the intended token in metadata.
+            string nameMaybe = null;
+            foreach (var argument in decodedValue.NamedArguments)
+            {
+                if (argument.Name == "Name")
+                {
+                    nameMaybe = (string)argument.Value;
+                }
+            }
+
+            // If the Name isn't defined, then use the name of the method.
+            if (nameMaybe != null)
+            {
+                name = nameMaybe;
+            }
+            else
+            {
+                // The Constructor case has an implied value provided by
+                // the runtime. We are going to enforce this during consumption
+                // so we avoid the setting of the value. We validate the name
+                // as empty at the use site.
+                if (kind is not UnsafeAccessorKind.Constructor)
+                {
+                    name = nameMaybe;
+                }
+            }
+
+            return true;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct GenerationContext
+        {
+            public UnsafeAccessorKind Kind;
+            public EcmaMethod Declaration;
+            public TypeDesc TargetType;
+            public bool IsTargetStatic;
+            public MethodDesc TargetMethod;
+            public FieldDesc TargetField;
+        }
+
+        private static TypeDesc ValidateTargetType(TypeDesc targetTypeMaybe)
+        {
+            TypeDesc targetType = targetTypeMaybe.IsByRef
+                ? ((ParameterizedType)targetTypeMaybe).ParameterType
+                : targetTypeMaybe;
+
+            // Due to how some types degrade, we block on parameterized
+            // types. For example ref or pointer.
+            if ((targetType.IsParameterizedType && !targetType.IsArray)
+                || targetType.IsFunctionPointer)
+            {
+                ThrowHelper.ThrowBadImageFormatException();
+            }
+
+            return targetType;
+        }
+
+        private static bool DoesMethodMatchUnsafeAccessorDeclaration(ref GenerationContext context, MethodDesc method, bool ignoreCustomModifiers)
+        {
+            MethodSignature declSig = context.Declaration.Signature;
+            MethodSignature maybeSig = method.Signature;
+
+            // Check if we need to also validate custom modifiers.
+            // If we are, do it first.
+            if (!ignoreCustomModifiers)
+            {
+                // One signature has custom modifiers the other doesn't, no match.
+                if (declSig.HasEmbeddedSignatureData != maybeSig.HasEmbeddedSignatureData)
+                {
+                    return false;
+                }
+
```

Comment:
I suspect that this may produce different results from unmanaged impl in corner cases. We may want to try to construct some of them.

Reply:
I tried a few permutations, definitely not exhaustive. What was your thinking here? Since this is the precise match I'm not sure how the unmanaged impl will deviate when custom modifiers are required to match precisely. For example `stdcall` will only be set here if it is the only one and that is fine because it too is an `EmbeddedSignatureData` instance. Since this is the precise case I don't see how the two can deviate.

Reply:
Example:
```
using System.Runtime.CompilerServices;

unsafe class Test
{
    static void Main() { F(default(Test), null); }

    [UnsafeAccessorAttribute(UnsafeAccessorKind.StaticMethod)]
    extern static void F(Test @this, delegate* unmanaged[Stdcall, MemberFunction]<void> p);

    static void F(delegate* unmanaged[Cdecl]<void> p) { Console.WriteLine("Here"); }
}
```
This will throw MissingMethodException with JIT and pass with NAOT.

Reply:
> This will throw MissingMethodException with JIT and pass with NAOT.

The first question to ask: Which one is the correct behavior?

Reply:
Ah. The case where we validate the actual calling convention bit as opposed to the custom modifiers. In this case the correct approach is to fail because the calling convention bits don't match and that is part of the embedded signature. I thought I was checking for that in NAOT too...

Reply:
Sigh... This doesn't do what I thought it does.

```csharp
// Validate calling convention.
if ((declSig.Flags & MethodSignatureFlags.UnmanagedCallingConventionMask)
    != (maybeSig.Flags & MethodSignatureFlags.UnmanagedCallingConventionMask))
{
    return false;
}
```

Reply:
> In this case the correct approach is to fail because the calling convention bits

We generally treat unmanaged calling convention details as modopts, even when they are encoded using legacy encoding. We should do the same here for consistency. I think both of the following examples should succeed (right know, the first one fails and the second one succeeds):

```csharp
[UnsafeAccessorAttribute(UnsafeAccessorKind.StaticMethod)]
extern static void F(Test @this, delegate* unmanaged[Stdcall]<void> p);

static void F(delegate* unmanaged[Cdecl]<void> p) { }
```

and

```csharp
[UnsafeAccessorAttribute(UnsafeAccessorKind.StaticMethod)]
extern static void F(Test @this, delegate* unmanaged[Stdcall, MemberFunction]<void> p);

static void F(delegate* unmanaged[Cdecl, MemberFunction]<void> p) { }
```

Reply:
That is a fair argument. I'll implement that.